### PR TITLE
all: Support syncing extended attributes (fixes #2698)

### LINKED ISSUE #2698: Sync extended file attributes on supported file systems
For now, Syncthing is not preserve extended file attributes (xattrs) when sync, as says in FAQ. But some tagging engines for Linux (such as Baloo, former Nepomuk from KDE) stores metainformation (tags, comments, ratings etc.) exactly in xattrs, therefore this metainfo are not synchronized across the cluster. Maybe is possible to add optional feature to sync xattrs across nodes, if file system supports it?

Some (possibly) related articles:
https://en.wikipedia.org/wiki/Extended_file_attributes#Linux
https://community.kde.org/Baloo/Architecture

P.S. I apologise for my very bad english. Hope that you can understand me.

### PR DESCRIPTION
**Pull request by calmh**
This adds support for syncing extended attributes on supported
filesystems on Linux, macOS, FreeBSD and NetBSD. Windows is currently
excluded because the APIs seem onerous and annoying and frankly the uses
cases seem few and far between. On Unixes this also covers ACLs as those
are stored as extended attributes.

Similar to ownership syncing this will optional & opt-in, which two
settings controlling the main behavior: one to "sync" xattrs (read &
write) and another one to "scan" xattrs (only read them so other devices
can "sync" them, but not apply any locally).

There are also limits in place on the size of individual attributes and the sum total per file, and a filtering mechanism which can be configured advancedly in the config file...

Adds checkboxes to folder advanced section:

Screenshot 2022-08-25 at 10 38 48

### CODE REVIEW

**Comment by tomasz1986** [comment]
Just some ideas/questions regarding the GUI.

1. Wouldn't it be more fitting to call the other options `sendOwnership` and `sendXattrs` instead of `scanOwnership` and `scanXattrs`? The act of "sending" seems to be the main purpose of the two, not just scanning.

2. The GUI looks a little cluttered. Maybe divide the two into their own subcategories, e.g. "Ownership" and "Extended Attributes"? All still inside the Advanced tab, of course.

3. The two "sync" options should probably be disabled (greyed out) when the folder is set to Send Only. Also, what about Receive Encrypted?

File: gui/default/syncthing/folder/editFolderModalView.html
Comment (tomasz1986) [inline gui/default/syncthing/folder/editFolderModalView.html:301]: "Implied by" sounds a little vague, especially if you don't know the internals. How about "automatically enabled by" instead?

File: gui/default/syncthing/folder/editFolderModalView.html
Comment (tomasz1986) [inline gui/default/syncthing/folder/editFolderModalView.html:316]:
```suggestion
                Enables sending extended attributes to other devices, but not applying incoming extended attributes. This can have a significant performance impact. Implied by "Sync Extended Attributes".
```

**Comment by calmh** [comment]
@syncthing/maintainers

**Comment by tomasz1986** [comment]
How about adding headings and dividing the two in the GUI like this?

image

It's not really possible to mark the changes using the GitHub interface, as the help link elements need to be moved, but I'm attaching a patch below.

```
From 1018c77cfc0da94ebbef22dbcf16c0d580af739d Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Tomasz=20Wilczy=C5=84ski?= <twilczynski@naver.com>
Date: Sun, 4 Sep 2022 10:19:43 +0200
Subject: [PATCH] wip
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Signed-off-by: Tomasz Wilczyński <twilczynski@naver.com>
---
 gui/default/syncthing/folder/editFolderModalView.html | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/gui/default/syncthing/folder/editFolderModalView.html b/gui/default/syncthing/folder/editFolderModalView.html
index 73cea66ed..fcd383af5 100644
--- a/gui/default/syncthing/folder/editFolderModalView.html
+++ b/gui/default/syncthing/folder/editFolderModalView.html
@@ -287,10 +287,13 @@
 
           <div class="row">
             <div class="col-md-6 form-group">
+              <p>
+                <label translate>Ownership</label>
+                &nbsp;<a href="{{docsURL('advanced/folder-sync-ownership')}}" target="_blank"><span class="fas fa-question-circle"></span>&nbsp;<span translate>Help</span></a>
+              </p>
               <label>
                 <input type="checkbox" ng-disabled="currentFolder.type == 'sendonly' || currentFolder.type == 'receiveencrypted'" ng-model="currentFolder.syncOwnership" /> <span translate>Sync Ownership</span>
               </label>
-              &nbsp;<a href="{{docsURL('advanced/folder-sync-ownership')}}" target="_blank"><span class="fas fa-question-circle"></span>&nbsp;<span translate>Help</span></a>
               <p translate class="help-block">
                 Enables sending ownership information to other devices, and applying incoming ownership information. Typically requires running with elevated privileges.
               </p>
@@ -302,10 +305,13 @@
               </p>
             </div>
             <div class="col-md-6 form-group">
+              <p>
+                <label translate>Extended Attributes</label>
+                &nbsp;<a href="{{docsURL('advanced/folder-sync-xattrs')}}" target="_blank"><span class="fas fa-question-circle"></span>&nbsp;<span translate>Help</span></a>
+              </p>
               <label>
                 <input type="checkbox" ng-disabled="currentFolder.type == 'sendonly' || currentFolder.type == 'receiveencrypted'" ng-model="currentFolder.syncXattrs" /> <span translate>Sync Extended Attributes</span>
               </label>
-              &nbsp;<a href="{{docsURL('advanced/folder-sync-xattrs')}}" target="_blank"><span class="fas fa-question-circle"></span>&nbsp;<span translate>Help</span></a>
               <p translate class="help-block">
                 Enables sending extended attributes to other devices, and applying incoming extended attributes. May require running with elevated privileges.
               </p>
-- 
2.35.1.windows.2
```

[0001-wip.patch.txt](https://github.com/syncthing/syncthing/files/9484291/0001-wip.patch.txt)

For the record, the HTML isn't great, as the `p` elements should really be just headings and `label` shouldn't exist without a corresponding `input` element, but a) the heading hierarchy in the GUI is currently broken (i.e. there's no `h1` (except for one obscure place where it shouldn't be), no `h2`, only `h3` and `h4`), and while theoretically it would be possible to use `h5` here, the styling doesn't match the rest of the interface, and b) `label` is unfortunately already used in many places in the GUI in place of headings, so if this was to be fixed, then probably all of them should be changed into `h` tags at once, including a CSS fix, which doesn't really belong to this PR.

File: lib/fs/basicfs_xattr_unix.go
Comment (calmh) [reply lib/fs/basicfs_xattr_unix.go:123]: No, because we apply the same filter when loading "current" attributes, so the ones that don't match the filter we will simply not know about, and neither remove nor overwrite.

File: lib/fs/basicfs_xattr_unix.go
Comment (calmh) [reply lib/fs/basicfs_xattr_unix.go:59]: Hm, yes, potentially. But I think it's desirable that a single large attribute (say a preview of a photo) doesn't prevent syncing small ones (e.g. ACLs). I think the order of returned attributes will generally be deterministic in some manner (alphabetical or chronological perhaps) so the flip-flopping should ideally be minimal. We could sort them ourselves though...

Comment (AudriusButkevicius) [reply lib/fs/basicfs_xattr_unix.go:59]: I'd say its fair to do that.

**Comment by AudriusButkevicius** [comment]
So had a look, a lot of content, but looks sensible. Another pair of eyes would be useful, and might have a chance for a second deeper look sometime soon.

**Comment by calmh** [comment]
It's a bit too high level and only does user attributes on BSD. Though I see it has a couple of corner case fixes for things like SMB that I should adopt.

File: gui/default/syncthing/folder/editFolderModalView.html
Comment (tomasz1986) [inline gui/default/syncthing/folder/editFolderModalView.html:301]:
```suggestion
                <input type="checkbox" ng-checked="currentFolder.sendOwnership || currentFolder.syncOwnership" ng-disabled="currentFolder.type == 'receiveonly' || currentFolder.type == 'receiveencrypted' || currentFolder.syncOwnership" ng-model="currentFolder.sendOwnership" /> <span translate>Send Ownership</span>
```

File: gui/default/syncthing/folder/editFolderModalView.html
Comment (tomasz1986) [inline gui/default/syncthing/folder/editFolderModalView.html:319]:
```suggestion
                <input type="checkbox" ng-checked="currentFolder.sendXattrs || currentFolder.syncXattrs" ng-disabled="currentFolder.type == 'receiveonly' || currentFolder.type == 'receiveencrypted' || currentFolder.syncXattrs" ng-model="currentFolder.sendXattrs" /> <span translate>Send Extended Attributes</span>
```

File: gui/default/syncthing/folder/editFolderModalView.html
Comment (tomasz1986) [reply gui/default/syncthing/folder/editFolderModalView.html:301]: This makes the "send" checkbox disabled and ticked when "sync" is ticked. The first condition is needed when the user ticks "send" first, only then ticks "sync", and then unticks "sync" again, or otherwise "send" will automatically untick as well.

image

The same applies to the extended attributes part below.

File: lib/protocol/bep_extensions.go
Comment (imsodin) [inline lib/protocol/bep_extensions.go:279]: `xattrsEqual` could handle the nils for a bit less verbouse comparisons in this block.

File: lib/protocol/bep_extensions.go
Comment (imsodin) [inline lib/protocol/bep_extensions.go:278]: Should this really check all OSes? It is used to e.g. compare an ad hoc create fileinfo from disk with one from a remote. If that one accumulated multiple OSes it will never be equal, would it?

File: lib/protocol/bep_extensions.go
Comment (imsodin) [inline lib/protocol/bep_extensions.go:419]: To be removed?

File: lib/scanner/walk.go
Comment (imsodin) [inline lib/scanner/walk.go:77]: I'd prefer to drop the `Get`

File: lib/fs/basicfs_fileinfo_bsdish.go
Comment (imsodin) [inline lib/fs/basicfs_fileinfo_bsdish.go:8]: Nice filename for those :)

File: lib/fs/basicfs_fileinfo_linuxish.go
Comment (imsodin) [inline lib/fs/basicfs_fileinfo_linuxish.go:8]: And openbsd is more linuxish than bsdish - do you think we can withstand the shitstorm if die-hard-bsdler discovery this (never encountered them) :P

File: lib/fs/basicfs_xattr_bsdish.go
Comment (imsodin) [inline lib/fs/basicfs_xattr_bsdish.go:25]: What is this syntax? I only know `[...]` for arrays as below, but this is a int to string map.

Also make both a const outside of func?

File: lib/fs/basicfs_xattr_bsdish.go
Comment (imsodin) [inline lib/fs/basicfs_xattr_bsdish.go:36]: Wouldn't you want to a abort if <=? What are the chances of not getting the exact same error again if doing the call again with the same `buf`.

File: lib/fs/basicfs_xattr_bsdish.go
Comment (imsodin) [inline lib/fs/basicfs_xattr_bsdish.go:42]: I dislike `%q` for logging/errors - an escaped path just looks bad in my opinion. What's the motivation for using it?

File: lib/fs/basicfs_xattr_bsdish.go
Comment (imsodin) [inline lib/fs/basicfs_xattr_bsdish.go:49]:
```suggestion
		// attribute name is not terminated by ASCII 0 (nul)."
```

File: lib/fs/basicfs_xattr_bsdish.go
Comment (imsodin) [inline lib/fs/basicfs_xattr_bsdish.go:63]: We could just skip appending if `l == 0` above instead

File: lib/fs/basicfs_xattr_linuxish.go
Comment (imsodin) [inline lib/fs/basicfs_xattr_linuxish.go:34]: Same here about why we retry if size <= len(buf)

File: lib/fs/basicfs_xattr_unix.go
Comment (imsodin) [inline lib/fs/basicfs_xattr_unix.go:34]:
```suggestion
	res := make([]protocol.Xattr, 0, len(attrs))
```

File: lib/fs/basicfs_xattr_unix.go
Comment (imsodin) [inline lib/fs/basicfs_xattr_unix.go:67]: This is redundant, as `attrs` is already sorted in the list attrs func.

**Review by imsodin — COMMENTED** [review]
Finally had a free night where I wasn't already dead tired to look through this.  
Most comments are nits that I noticed while getting to know what's going on. I have one more fundamental comment, which likely is a misconception (will link after posting) - once that's cleared up or addressed I'll need to go over the model/folders again.

Have you considered https://github.com/pkg/xattr?  
Looks like it's actively maintained. Not sure if it's better to have our own code or a public dependency even if it would fit our needs though (more overhead to do changes).

There's a few changes around owner related parameters and comparisons - would be nice to split them out (unless the ensuing conflict hell is too bad of course).

File: lib/model/folder_sendrecv.go
Comment (calmh) [reply lib/model/folder_sendrecv.go:1792]: The only thing we update here is the inode change time, which we need for comparisons in the scanner. We need to do it here because it's affected by things like chmod, chtimes and setting xattrs which we've done previously. That attribute is never sent over the wire so there are no discrepancies.

For the actual xattrs, we don't change what's in the global version based on our filtering etc when we apply. However, if the file is then changed locally in whatever way, our filtering will be applied when we scan it. I think this is as desired...

File: lib/protocol/bep_extensions.go
Comment (calmh) [reply lib/protocol/bep_extensions.go:278]: This outer `f.Platform != other.Platform` check is just a shortcut to catch the easy case where all platform values are equal or nil, then we don't need to do per-platform comparisons. If there is something unequal we'll do the inner `xattrsEqual` checks which return true if either side is nil (nothing to compare to). That is, things will only be considered *unequal* iff there is data for a given platform on both sides and it differs.

A fileinfo created on Linux will never be unequal to another just because the other has Windows attributes, for example -- it will only be unequal if there are also Linux attributes and those differ.

File: lib/protocol/bep_extensions.go
Comment (calmh) [reply lib/protocol/bep_extensions.go:279]: (done)

File: lib/model/folder_sendrecv.go
Comment (calmh) [reply lib/model/folder_sendrecv.go:595]: Hmm, maybe we're not entirely consistent here, but I think praxis is for the prefix to be that of the executing function. That is, if I call `os.Open("foo")` it will return `open: foo: no such file`, so I took `handling dir` as we are in `handleDir`... The outer context will be why we're handling the dir, maybe.

File: lib/scanner/walk.go
Comment (calmh) [reply lib/scanner/walk.go:77]: So would I, but the attributes are called MaxSingleEntrySize etc. so when interface:ing we need to differentiate

File: lib/fs/basicfs_fileinfo_linuxish.go
Comment (calmh) [reply lib/fs/basicfs_fileinfo_linuxish.go:8]: I will challenge them to an office jousting competition

File: lib/fs/basicfs_xattr_bsdish.go
Comment (calmh) [reply lib/fs/basicfs_xattr_bsdish.go:25]: > but this is a int to string map.

Haha! So you might think, but then I fooled you. This is in fact an array constant with values of the form `offset: value`.

But yeah, this could be moved outside.

File: lib/fs/basicfs_xattr_bsdish.go
Comment (calmh) [reply lib/fs/basicfs_xattr_bsdish.go:36]: So there's a detail in how the syscall works here. If called with a non-zero buffer that isn't big enough, it will return the number of bytes written (== the buffer size) and ERANGE. However if we call it with a zero sized buffer (nil, in this case) it returns the number of bytes actually required. In principle I guess we could skip the `if` and always do the `make` because it should always return a size that's bigger than the buffer we already had...

File: lib/fs/basicfs_xattr_bsdish.go
Comment (calmh) [reply lib/fs/basicfs_xattr_bsdish.go:42]: ... I don't recall, and I agree. Will change.

**Comment by calmh** [comment]
@imsodin PTAL :)

File: lib/model/folder_sendrecv.go
Comment (imsodin) [reply lib/model/folder_sendrecv.go:595]: I always understood those contexts to differentiate multiple errors returned from the same function, so one can know from which operation inside the function came from. Whichever it is we should conform to one, otherwise we end up with "handle dir: handling dir: err" at some point. Anyway, not a concern for this PR.

File: lib/model/folder_sendrecv.go
Comment (imsodin) [inline lib/model/folder_sendrecv.go:613]: Actually the conflict move context only applies to the error on L606, L610 just deletes stuff. Maybe just drop that context in this PR.

File: lib/model/folder_sendrecv.go
Comment (imsodin) [reply lib/model/folder_sendrecv.go:1792]: Right. Two suggestions to make that more understandable (to me at least :) ):

Add a comment, e.g. "Operations while pulling might have changed inode-time. The scanner uses that to detect changes. Thus the file info is update such that the subsequent scan doesn't pick it up."

Metadata is just inode-change set. xattrs, owner, group, perms, ... are something else than metadata. That's not obvious to me at least. I'd be in favor of renaming it to e.g. `fileInfoUpdateInodeTime`. If there is more metadata in the future, we can always create a generic function again and call the inode-time one from there.

File: lib/model/folder_sendrecv.go
Comment (calmh) [reply lib/model/folder_sendrecv.go:595]: I take it upon myself to harmonize errors sanely at some point.

**Review by imsodin — DISMISSED** [review]

**Review by imsodin — APPROVED** [review]

**Comment by acolomb** [comment]
@calmh Please don't forget updating the docs: API (platform and inodeChange in example), config (xattrs, scan --> send), and protocol.

**Pull request merged**
